ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs

```diff
@@ -313,3 +378,105 @@ prop_appendOldStateRestoresPreWindow TestChainState{..} =
 
 size :: Num b => Seq.StrictSeq a -> b
 size = fromIntegral . Seq.length
+
+data Input c
+  = InputEBB !(WithOrigin SlotNo) !SlotNo
+    -- ^ the preceding signed slot, the EBB's slot
+  | InputSigner !(CS.PBftSigner c)
+  deriving (Eq, Show)
+
+newtype Inputs c = Inputs {unInputs :: [Input c]}
+  deriving (Eq, Show)
+
+signatureInputs :: Inputs c -> [CS.PBftSigner c]
+signatureInputs (Inputs inps) = [ signer | InputSigner signer <- inps ]
+
+ebbInputs :: Inputs c -> [(WithOrigin SlotNo, SlotNo)]
+ebbInputs (Inputs inps) = [ (prev, slot) | InputEBB prev slot <- inps ]
+
+slotInput :: Input c -> SlotNo
+slotInput = \case
+  InputEBB _ slot -> slot
+  InputSigner   x -> CS.pbftSignerSlotNo x
+
+slotInputs :: Inputs c -> [SlotNo]
+slotInputs = map slotInput . unInputs
+
+lengthInputs :: Num a => Inputs c -> a
+lengthInputs = fromIntegral . length . unInputs
+
+headSigner :: Inputs c -> CS.PBftSigner c
+headSigner = head . signatureInputs
+
+safeLastSigner :: Inputs c -> Maybe (CS.PBftSigner c)
+safeLastSigner = listToMaybe . reverse . signatureInputs
+
+appendInput ::
+     PBftCrypto c
+  => SecurityParam
+  -> CS.WindowSize
+  -> Input c
+  -> CS.PBftChainState c -> CS.PBftChainState c
+appendInput k n = \case
+  InputEBB _ slot    -> CS.appendEBB k n slot
+  InputSigner signer -> CS.append k n signer
+
+appendInputs ::
+     PBftCrypto c
+  => SecurityParam
+  -> CS.WindowSize
+  -> Inputs c
+  -> CS.PBftChainState c -> CS.PBftChainState c
+appendInputs k n = repeatedly (appendInput k n) . unInputs
+
+-- | May include EBBs before the first signer and after the last signer
+takeSigners :: Word64 -> Inputs c -> Inputs c
+takeSigners = \n0 -> Inputs . go n0 . unInputs
+  where
+    go !n = \case
```

Comment:
(Curious about the bang: since we compare it against 0 before it can become a thunk (`n - 1`), it will either be in normal form (`v` for value) or `v - 1`. You can ignore this comment.)

Reply:
I added that by habit: since the `[]` case doesn't use `n`, GHC would not infer that the `go` function is strict, so the WorkerWrapper optimization wouldn't fire to unbox the `n` argument, yadda yadda loop yadda yadda.